scripts/sell_unused_slots.py

```diff
@@ -0,0 +1,112 @@
+from datetime import (
+    datetime,
+    timedelta,
+)
+import sys
+from typing import (
+    Iterable,
+)
+
+from azul import (
+    cached_property,
+    logging,
+)
+from azul.bigquery_reservation import (
+    SlotManager,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.logging import (
+    configure_script_logging,
+)
+
+log = logging.getLogger(__name__)
+
+
+class ReindexMonitor:
+    # Minutes
+    interval = 5
+
+    # Minimum number of `-contribute` Lambda invocations per interval for a
+    # reindexing to be considered active
+    threshold = 0
+
+    @cached_property
+    def _cloudwatch(self):
+        return aws.client('cloudwatch')
+
+    @cached_property
+    def _lambda(self):
+        return aws.client('lambda')
+
+    def is_reindex_active(self) -> bool:
+        for function_name in self.list_contribution_lambdas():
+            invocation_count = self._lambda_invocation_count(function_name)
+            logging_args = (function_name, f'{invocation_count} invocations in the last '
+                                           f'{self.interval} minutes')
+            if invocation_count > self.threshold:
+                log.info('Detected active reindex for Lambda %r; %s', *logging_args)
+                return True
+            else:
+                log.debug('Lambda %r is inactive (%s)', *logging_args)
+        return False
+
+    def list_contribution_lambdas(self) -> Iterable[str]:
+        """
+        Search Lambda functions for the names of contribution Lambdas.
+
+        This method retrieves over 350 kilobytes of data from AWS if its output
+        is greedily consumed.
+        """
+        # The default and maximum number of items returned is 50.
+        # Reducing this number will result in more round trips being made
+        # here when no reindex is active, but will reduce the cumulative
+        # response size if an active reindex is found early in the loop.
+        kwargs = dict(MaxItems=20)
+        while True:
+            response = self._lambda.list_functions(**kwargs)
+            for function in response['Functions']:
+                function_name = function['FunctionName']
+                if function_name.endswith('-contribute'):
+                    yield function_name
```

Comment:
I wanted to be able to quit the loop early due to the amount of data being retrieved but since that's not an actual concern there's no remaining reason.